### PR TITLE
[MI-2369]:Created edit and get subscription template API

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**

### CODE REVIEW

File: server/subscribe.go [hunk 1]
```diff
@@ -1105,24 +1152,58 @@ func (p *Plugin) httpGetSubscriptionTemplates(w http.ResponseWriter, r *http.Req
...
+	return respondJSON(w, subTemplates)
+}
+
+func (p *Plugin) httpEditSubscriptionTemplates(w http.ResponseWriter, r *http.Request, mattermostUserID string) (int, error) {
+	subscriptionTemplate := SubscriptionTemplate{}
+	err := json.NewDecoder(r.Body).Decode(&subscriptionTemplate)
+	if err != nil {
+		return respondErr(w, http.StatusBadRequest,
+			errors.WithMessage(err, "failed to decode the incoming request"))
+	}
```
Comment (ayusht2810) [inline server/subscribe.go:1190]:
```suggestion
	if err := json.NewDecoder(r.Body).Decode(&subscriptionTemplate); err != nil {
		return respondErr(w, http.StatusBadRequest,
			errors.WithMessage(err, "failed to decode the incoming request"))
	}
```

File: server/subscribe.go [hunk 2]
```diff
@@ -1105,24 +1152,58 @@ func (p *Plugin) httpGetSubscriptionTemplates(w http.ResponseWriter, r *http.Req
 			errors.Wrap(err, "unable to get channel subscription templates"))
 	}
 
-	client, _, _, err := p.getClient(instanceID, types.ID(mattermostUserID))
+	var subTemplates []SubscriptionTemplate
+
+	projectKey := r.FormValue("project_key")
```
Comment (ayusht2810) [inline server/subscribe.go:1157]: constant for `project_key`

File: server/subscribe.go [hunk 3]
```diff
@@ -1105,24 +1152,58 @@ func (p *Plugin) httpGetSubscriptionTemplates(w http.ResponseWriter, r *http.Req
...
+	if len(projectKey) < 1 {
+		client, _, _, err := p.getClient(instanceID, types.ID(mattermostUserID))
+		if err != nil {
+			return respondErr(w, http.StatusInternalServerError, err)
+		}
+
+		plist, err := client.ListProjects("", -1)
```
Comment (ayusht2810) [inline server/subscribe.go:1164]:
```suggestion
		pList, err := client.ListProjects("", -1)
```

File: server/subscribe.go [hunk 4]
```diff
@@ -1105,24 +1152,58 @@ func (p *Plugin) httpGetSubscriptionTemplates(w http.ResponseWriter, r *http.Req
 			errors.Wrap(err, "unable to get channel subscription templates"))
 	}
 
-	client, _, _, err := p.getClient(instanceID, types.ID(mattermostUserID))
+	var subTemplates []SubscriptionTemplate
```
Comment (ayusht2810) [inline server/subscribe.go:1155]:
```suggestion
	var subTemplates []*SubscriptionTemplate
```

**Review by ayusht2810 — APPROVED** [review]

File: server/subscribe.go [hunk 5]
```diff
@@ -377,6 +387,45 @@ func (p *Plugin) addSubscriptionTemplate(instanceID types.ID, newSubscriptionTem
...
+	subKey := keyWithInstanceID(instanceID, templateKey)
+	return p.atomicModify(subKey, func(initialBytes []byte) ([]byte, error) {
+		subscriptionTemplates, err := SubscriptionTemplatesFromJSON(initialBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		oldSubscriptionTemplate, ok := subscriptionTemplates.Templates.ByID[modifiedSubscriptionTemplate.ID]
+		if !ok {
+			return nil, errors.New("existing subscription template does not exist")
```
Comment (manojmalik20) [inline server/subscribe.go:400]:
```suggestion
			return nil, errors.New("subscription template does not exist")
```

File: server/subscribe.go [hunk 6]
```diff
@@ -377,6 +387,45 @@ func (p *Plugin) addSubscriptionTemplate(instanceID types.ID, newSubscriptionTem
...
+		if modifiedSubscriptionTemplate.Filters.Projects.Len() == 1 {
+			newProjectKey = modifiedSubscriptionTemplate.Filters.Projects.Elems()[0]
+		}
+
+		if err = p.validateSubscriptionTemplate(modifiedSubscriptionTemplate, instanceID, client, newProjectKey); err != nil {
+			return nil, err
+		}
+
+		subscriptionTemplates.Templates.delete(oldProjectKey, oldSubscriptionTemplate.ID)
+		subscriptionTemplates.Templates.add(modifiedSubscriptionTemplate, newProjectKey)
```
Comment (manojmalik20) [inline server/subscribe.go:418]: It would be great if the `projectKey` is the first parameter in the `add` function.

File: server/subscribe.go [hunk 7]
```diff
@@ -1105,24 +1154,57 @@ func (p *Plugin) httpGetSubscriptionTemplates(w http.ResponseWriter, r *http.Req
...
+		}
+	}
+	return respondJSON(w, subTemplates)
+}
+
+func (p *Plugin) httpEditSubscriptionTemplates(w http.ResponseWriter, r *http.Request, mattermostUserID string) (int, error) {
+	subscriptionTemplate := SubscriptionTemplate{}
+	if err := json.NewDecoder(r.Body).Decode(&subscriptionTemplate); err != nil {
+		return respondErr(w, http.StatusBadRequest,
+			errors.WithMessage(err, "failed to decode the incoming request"))
```
Comment (manojmalik20) [inline server/subscribe.go:1190]:
```suggestion
		return respondErr(w, http.StatusBadRequest, errors.WithMessage(err, "failed to decode the incoming request"))
```

**Review by manojmalik20 — COMMENTED** [review]
Please resolve the conflicts.

File: server/subscribe.go [hunk 8]
```diff
@@ -477,9 +480,9 @@ func (p *Plugin) validateSubscriptionTemplate(subscriptionTemplate *Subscription
 		return err
 	}
 
-	for _, subscriptionTemplate := range templates.Templates.ByProjectID[projectKey] {
-		if subscriptionTemplate.Name == subscription.Name && subscriptionTemplate.ID != subscription.ID {
-			return errors.Errorf("Subscription name, '%s', already exists. Please choose another name.", subscription.Name)
+	for _, template := range templates.Templates.ByProjectID[projectKey] {
+		if template.Name == subscriptionTemplate.Name {
```
Comment (manojmalik20) [inline server/subscribe.go:484]: Why did you remove the condition of checking the IDs?

**Review by manojmalik20 — APPROVED** [review]
